Datacenter simulation (mostly) works between two nodes.

randofan opened the pull request:
The simulation runs between two nodes. It sends traffic between the two nodes in both directions. Flows are scheduled based on traffic generation outlined in `traffic_gen/README.md`.

Changes to `src/`:
1. I added a trace on m_dataRetries inside of `tcp-socket-base.h` to track retransmissions.
2. I added a max SACK field to `tcp-rx-buffer.h` to limit the size of the SACK list in ACKs.
3. I changed `BulkSendApplication::SendData` to only send data when the function is explicitly called by a user. I disabled the BulkSendApplication from immediately sending bytes once the connection has succeeded. I changed how `m_totbytes` and `m_maxBytes` operates. Now, `m_totBytes` represents the total bytes so far sent in the flow and `m_maxBytes` is the size of the flow. `BulkSendApplication::SendData` sends bytes until it's finished the flow.

Issues:
1. I need to re-work how I analyze FCT based on the changes in traffic generation from before. I wasn't able to find anything conclusive about how the High Precision repository measures FCT. I'll prioritize getting this set up first.
3. the "Scalable Tail Latency Estimation for Data Center Networks" describes that packet loss is extremely rare in datacenters and most errors come from over-subscription. The [High Precision repository](https://github.com/alibaba-edu/High-Precision-Congestion-Control) also has a 0 link error rate. I'm not too sure how to accurately model this with my topology and flows.
4. I wasn't able to read the traffic_gen directly from `scratch/tmp_traffic.txt`. I'm currently just using a fixed array of strings, which represents each line from the file. I'm not too sure what I'm doing wrong with my file read. I've commented out my attempt inside of `scratch/simulation.cc`. It'd be great if you could take a look at it.
5. Due to the modifications to BulkSendApplication, it isn't compatible with previous implementations. It might be better to create a new Application. However, I'm unsure what issues could arise from adding a new Application (such as adding it to the import module).

Commit:
fixed issue with build fail. now, socket won't connect

Commit:
i don't know what is wrong. the tcp socket won't connect correctly & now I get a <Signals.SIGSEGV: 11>

Commit:
can't connect socket

Commit:
last minute changes

Commit:
still doesn't work

Commit:
working version

Commit:
trace won't connect on tcpsocketbase

Commit:
sequence number and highest ack work, retransmit doesn't

Commit:
able to send flow based on traffic_gen between two nodes

rajathshashidhara commented on scratch/simulation.cc:
```diff
@@ -0,0 +1,309 @@
+
+    // Set error rate
+    Ptr<UniformRandomVariable> uv = CreateObject<UniformRandomVariable>();
+    uv->SetStream(50);
```
What does this parameter control?

randofan replied:
I think its a way to make the uniform random variable deterministic. The documentation says that it's randomly assigned by default.

rajathshashidhara commented on scratch/simulation.cc:
```diff
@@ -0,0 +1,309 @@
+    em->SetRandomVariable(uv);
+    em->SetRate(error_rate);
+    em->SetUnit(RateErrorModel::ERROR_UNIT_PACKET);
+    p2p.SetDeviceAttribute("ReceiveErrorModel", PointerValue(em));
```
Does the same error model apply to both directions of the channel?

rajathshashidhara commented on scratch/simulation.cc:
```diff
@@ -0,0 +1,309 @@
+    // 
+
+    // std::ifstream file;
+    // file.open("tmp_traffic.txt");
```
Might be an issue due to use of relative path. Have you tried using the absolute path?

randofan replied:
```
std::ifstream file;
file.open("~/home/david/ns-3/scratch/tmp_traffic.txt");
NS_ASSERT(!file.is_open());
NS_ASSERT(!file.good());
NS_ASSERT(file.eof());
```

I've tried it with absolute path, and in both cases, the third assertion fails; meaning that it's able to open the file,  but it immediately sees and eof, which is definitely not the case. The same problem exists even if I give it a different file to read.

rajathshashidhara replied:
Try seeking to the beginning of the file. Maybe the file open() mode is incorrect.

randofan replied:
I did some more debugging, and I found its a issue with ns3 build system. Whenever I compile and run with gcc, the file stream works, but it doesn't work when I build with ns3. I've tried cleaning and re-building, and the problem persists. I think it may be something with CMake?

rajathshashidhara commented on src/applications/model/bulk-send-application.cc:
```diff
@@ -208,20 +240,18 @@ BulkSendApplication::StopApplication() // Called at time specified by Stop
-// Private helpers
-
 void
 BulkSendApplication::SendData(const Address& from, const Address& to)
```
is the idea here to transmit m_maxBytes each time this function is called to simulate a flow? If yes, why do we expect the application the produce flows of the same size? Do you plan to schedule calls to this function to simulate flows?

randofan replied:
The way I'm doing it is, for each generated flow, I'd schedule a callback at the given time. Inside the callback, I first set the m_maxBytes variable with the desired flow size, and then call SendData. I've since changed it to create a new BulkSendApplication::SendSize(uint32_t size) to send the desired number of bytes whenever its called.

rajathshashidhara commented:
To simulate packet loss due to over-subscription, we can under-provision the bandwidth in the core of the network.

randofan commented:
> To simulate packet loss due to over-subscription, we can under-provision the bandwidth in the core of the network.

How should I do this? The flow generation tool has two options 1) to change the percent of network load and 2) change the host bandwidth. Which should I try increasing?

randofan commented:
I'm also encountering an issue with the traces.
```
/**
 * Sender
**/
static void
SendTrace(Ptr<OutputStreamWrapper> stream, std::string context, SequenceNumber32 oldValue, SequenceNumber32 newValue)
{   
    std::cout << context << " Send " << std::to_string(Simulator::Now().GetSeconds()) << " " << newValue << std::endl;
    *stream->GetStream() << context << " Send " << std::to_string(Simulator::Now().GetSeconds()) << " " << newValue << std::endl;
}
```

Here's the callback I use for traces. Whenever I give it flows where the sender node switches between flows, the file stream doesn't trace them all, but the `std::cout` does. For example, given this input; I'm sending 200 bytes from node 1 to node 0 at 2.5 seconds and then another 200 bytes from node 0 to 1 at 3 seconds.
```
std::string input_data[] = {
  "1 0 3 100 200 2.500",  
  "0 1 3 100 200 3.000"
};
```
Here's what I see printed to stdout:
```
1 Send 2.500000 201
1 Recv 2.500002 201
0 Send 3.000000 201
0 Recv 3.000002 201
```
Here's what I see printed to the file designated by `stream`: 
```
0 Send 3.000000 201
0 Recv 3.000002 201
```
I've tested it a few time with different inputs and it seems the file stream misses printing some traces whenever I switch whoever is sending. The reason I'm so confused is because it works perfectly when I print to stdout, which leads me to believe is a problem with the file stream? Do you have any guidance on this problem?

rajathshashidhara commented:
> How should I do this? The flow generation tool has two options 1) to change the percent of network load and 2) change the host bandwidth. Which should I try increasing?

When you say network load, do you mean the network utilization? What is the topology you are using? What is the bandwidth of the core switch?

> I'm also encountering an issue with the traces.

Verify that `*stream->GetStream()` is valid when tracing either direction. I can imagine that it might only be set for Node 0.

randofan commented:
> When you say network load, do you mean the network utilization? What is the topology you are using? What is the bandwidth of the core switch?

The documentation just says "the percentage of the traffic load to the network capacity, by default 0.3". Right now, I'm still just using the two nodes sending to each other, but I'm working on implementing the fat tree topology.

> Verify that `*stream->GetStream()` is valid when tracing either direction. I can imagine that it might only be set for Node 0.

If I only schedule a flow from node 1, the trace records it. I think it might be an issue with two nodes concurrently trying to write to the file. Should I just make a different trace for each node?

rajathshashidhara commented:
- For a 2-node topology, oversubscription does not make sense. It would be great if you can define the network load more precisely in our next meeting -- is it the average load with possible spikes or is the network is constantly loaded at  that value?
- Are you running on multiple cores? If yes, you could try to run with just one core to see if that solves the problem. If using multiple files makes it easy for you, that's okay. Eventually, you should figure out a less hacky solution for logging -- it should be pretty standard in ns-3.

Commit:
added fat tree topology, can read from files, changed sendSize() to make it more intuitive

randofan commented:
- I was able to fix the issue with reading from files: it was a problem with the path as you predicted.
- I changed the SendData() to a SendSize(uint32_t size) function containing the number of bytes the client wants to send. I call this instead of SendData.
- I'm using a fat tree topology now. I found a helper class online, which I added to `src/`. Currently, the simulation doesn't run because I'm still trying to figure out how the helper class does IP addressing.
- I looked more into how `traffic_gen.py` uses `--load` and it seems that it uses it to calculate the average time between flows by multiplying it with the bandwidth. It then puts it into a poisson distribution when generating the flows themselves.
- I'm still trying to figure out how to trace the flows in a single file.
- I'm still working on the analysis; it seems pretty straightforward though.

Commit:
simulation now works

Commit:
working on final

Commit:
seqNum not summing up

Commit:
added traces

randofan commented:
I've been running the analysis on smaller numbers of flows to test it. It worked fine for 20 flows, but breaks when I upped it to 900 flows in a file called `small_traffic.txt`. 

I've realized the problem is that the sequence numbers are getting reset to 1 during the simulation. An example of this is for node 14. I've pushed the input flows in `14in.txt` and the output traces `14out.txt`. You can see that the sequence number is being set back to 1 on line 120 of `14out.txt`. These aren't retransmissions because 14 is getting an ack 537 beforehand and we're also well below integer overflow.

I found that all the sequence number resets are when a new flow is scheduled; there's a noticeable gap from when the message was sent for the last flow, and when the new flow starts (as seen in line 119-120 in `14out.txt`). However, afterwards, the two flows are then intertwined. You can see on line 131 that 14 sends the previous flow with the higher sequence number before going back to the new flow on line 162. 

This behavior is further not consistent across all flows. I'd previously seen the case that when a new flow was scheduled while a current flow was still in progress, it'd get queued and then would instead begin after the previous flow had finished. Thus, I wrote my analysis to merge the two flows into one large flow.

I didn't anticipate these two different behaviors for seemingly the same issue initially and this completely messes up the way I was tracking fct. The way I'm currently doing it is whenever I see a send trace whose time corresponded to when the next flow from that node should be sent, I record the sequence number and the size of the flow. Then, when I see an ack that's larger than the seqNum + size, I record the difference between send time and ack receive time as the fct.

I have no idea when TCP decides to follow which behavior. Understanding and let alone accounting for this behavior would definitely take at least a week, and it wouldn't be possible to complete by Wednesday, when I aim to turn in the report to Simon by.

Running it right now, I'm able to get roughly 150/900 flows where this case doesn't occur. Assuming this ratio stays roughly the same for the larger set of 3.4k flows, then we'd yield around 550 flows from that. And I'm not even too sure on how accurate those flows would be because they could be longer than they're supposed to be. If the sequence number was reset halfway through the flow, it'd cause the fct to be 1.5x its actual amount.

I was hoping for your advice on where to go from here?

Commit:
readability changes to compute-fct
